litex/build/xilinx/symbiflow.py

```diff
@@ -0,0 +1,321 @@
+                ]),
+            Rule("$(TOP).bit", ["$(TOP).fasm"], commands=[
+                    "write_bitstream -d $(BITSTREAM_DEVICE) -f $(TOP).fasm -p $(PARTNAME) -b $(TOP).bit > /dev/null"
+                ]),
+            Rule("clean", phony=True, commands=[
+                    "rm -f $(ARTIFACTS)"
+                ]),
+        ])
+
+        tools.write_to_file("Makefile", makefile.generate())
+
+    def _build_clock_constraints(self, platform):
+        for clk, (period, phase) in sorted(self.clocks.items(), key=lambda x: x[0].duid):
+            rising_edge = math.floor(period/360.0 * phase * 1e3)/1e3
+            falling_edge = math.floor(((rising_edge + period/2) % period) * 1.e3)/1e3
+            platform.add_platform_command(f"create_clock -period {period} {{clk}} -waveform {{{{{rising_edge} {falling_edge}}}}}", clk=clk)
+        for from_, to in sorted(self.false_paths, key=lambda x: (x[0].duid, x[1].duid)):
+           platform.add_platform_command("set_clock_groups -exclusive -group {{{from_}}} -group {{{to}}}", from_=from_, to=to)
+        # Make sure add_*_constraint cannot be used again
+        del self.clocks
+        del self.false_paths
+
+    # Yosys has limited support for real type. It requires that some values be multiplied
+    # by 1000 and passed as integers. For details, see:
+    # https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/xc/xc7/techmap/cells_map.v
+    def _fix_instance(self, instance):
+        if instance.of == "PLLE2_ADV":
+            for item in instance.items:
+                if isinstance(item, Instance.Parameter) and re.fullmatch("CLKOUT[0-9]_(PHASE|DUTY_CYCLE)", item.name):
+                    item.value = wrap(math.floor(_unwrap(item.value) * 1000))
+
+    def build(self, platform, fragment,
+        build_dir  = "build",
+        build_name = "top",
+        run        = True,
+        enable_xpm = False,
+        **kwargs):
+
+        self._check_properties(platform)
+
+        # Create build directory
+        os.makedirs(build_dir, exist_ok=True)
+        cwd = os.getcwd()
+        os.chdir(build_dir)
+
+        # Finalize design
+        if not isinstance(fragment, _Fragment):
+            fragment = fragment.get_fragment()
+        platform.finalize(fragment)
+
+        # Symbiflow-specific fixes
+        for instance in fragment.specials:
+            if isinstance(instance, Instance):
+                self._fix_instance(instance)
+
+        # Generate timing constraints
+        self._build_clock_constraints(platform)
+
+        # Generate verilog
+        v_output = platform.get_verilog(fragment, name=build_name, **kwargs)
+        named_sc, named_pc = platform.resolve_signals(v_output.ns)
+        v_file = build_name + ".v"
+        v_output.write(v_file)
+        platform.add_source(v_file)
+
+        self._generate_makefile(
+            platform   = platform,
+            build_name = build_name
+        )
+
+        # Generate design constraints
```

Comment:
Why do we need all of an `.xdc`, `.sdc` and `.pcf`?

Reply:
.pcf (list of pin locations) is required by `place` script.
.sdc (timing constraints) is optional. I'll change this so that this file will be generated and passed to the toolchain only when there are constraints set.
.xdc (IO settings) is optional. It is never empty, so it is easier to just generate and use it, even when this is not strictly needed. Is there any advantage in not using it?